### ISSUE TITLE
LegalMoveGenerator can run generate_legal_moves() twice when creating a list

### ISSUE THREAD

**Issue by mdoege** [niklasf/python-chess#226]
Python-chess is great!

But I have found that

`list(board.legal_moves)`

or

`[*board.legal_moves]`

to get a list is unnecessarily slow because generate_legal_moves() is called twice. (First from `__len__` of the LegalMoveGenerator, and then from `__iter__`.)

`[x for x in board.legal_moves]`

on the other hand is much faster because this double calling does not happen.

It took me quite a bit of profiling with SnakeViz and stack traces to figure this out, so maybe LegalMoveGenerator should get a to_list() method that returns a list in the fastest way possible.

**Comment by niklasf** [comment]
Wow! Great find! I added a test case to find out if all Python versions do this.

**Comment by PedanticHacker** [comment]
def generate_legal_moves(self):
    self.traversals += 1
    return
    yield

This is the test-case function just commited. Isn't everything after `return` considered dead code? Will `yield` ever be tested?

**Comment by PedanticHacker** [comment]
```
>>> def test():
...     a = 5
...     b = 10
...     return a + b
...     print("I am dead!")
...
>>> test()
15
```

The string `I am dead!` does not display, because everything after the `return` statement is dead code.

**Comment by niklasf** [comment]
Adding `to_list()` is a good idea to keep backwards compability, but I'd rather break compability to get rid of this trap. Instead I renamed `__len__` to `count` .

Thanks again for reporting this. This was the easiest 100% speedup in a while 😅. Also `len(list(self))` is now faster than the old `__len__()` because iterating in C is faster than in Python, adding another 2% perft speed (even in code that wasn't using list conversions originally).

**Comment by mdoege** [comment]
Great, thanks everyone for the fast response!

**Comment by niklasf** [comment]
Released in v0.22.0.
